File: rust/src/ray.rs
use crate::color::Color;
use crate::point3::Point3;
use crate::vec3::Vec3;

#[derive(Debug)]
pub struct Ray {
    pub origin: Point3,
    pub direction: Vec3,
    pub color: Color,
}

impl Ray {
    pub fn at(&self, t: f32) -> Point3 {
        self.origin + t * self.direction
    }
}

File: zig/src/ray.zig
const Vec3 = @import("vec3.zig").Vec3;
const Point3 = @import("point3.zig").Point3;

pub const Ray = struct {
    origin: Point3 = Point3{},
    direction: Vec3 = Vec3{},

    pub fn at(self: Ray, t: f32) Point3 {
        return self.origin + self.direction.mul(t);
    }
};